Basic descriptions track support

OwenEdwards opened the pull request:
Adds support for descriptions text tracks, with a separate menu button on the control bar, and display using the same mechanism as captions and subtitles. To handle the situation where a descriptions and a captions or subtitles track could be enabled at the same time, the TextTrackDisplay prefers non-descriptions tracks over descriptions tracks, and the descriptions menu button disables itself if another track is enabled.

This PR could probably use some more Unit tests to check the priority model is TextTrackDisplay, and the self-disabling in descriptions-button. Also, the CSS styling applied to the button when it is disabled (just making it 0.5 opacity) may need more work.

I'd appreciate any feedback, suggestions, etc.

gkatsev commented on src/css/components/_layout.scss:
```diff
@@ -124,6 +124,10 @@ body.vjs-full-window {
 // Hide disabled or unsupported controls.
 .vjs-hidden { display: none !important; }
 
+.vjs-disabled {
```
we can also change the [`cursor` `not-allowed`](https://developer.mozilla.org/en-US/docs/Web/CSS/cursor#Values)

OwenEdwards replied:
Wondering whether `cursor: not-allowed` is overkill, since a disabled button `<button type="button" disabled>Click Me!</button>` only shows `cursor: default` instead of `cursor: pointer`?

gkatsev replied:
`cursor: pointer` seems fine as well.

gkatsev commented on src/js/player.js:
```diff
@@ -2462,13 +2462,13 @@ class Player extends Component {
-    * Captions - text displayed over the video for the hearing impaired
-    * Subtitles - text displayed over the video for those who don't understand language in the video
-    * Chapters - text displayed in a menu allowing the user to jump to particular points (chapters) in the video
-    * Descriptions (not supported yet) - audio descriptions that are read back to the user by a screen reading device
```
:D

gkatsev commented:
Thanks @OwenEdwards, looks great.

OwenEdwards commented:
@gkatsev would love some help in building out the unit tests, as I mention. Otherwise, I think this functionality is ready to go with these latest CSS updates.

gkatsev commented:
Sorry it's been a while. I should be able to help out with tests for this PR this week.

gkatsev commented:
Also, looks like this needs to be rebase or updated from master.

OwenEdwards commented:
I'll take a look. If I do a rebase, then it'll have to become a new PR, right? Not a big deal, just wanted to know if I'm missing something.

gkatsev commented:
You can force push if you want. Or you could merge.

OwenEdwards commented:
Okay, this is rebased, simplified (fewer separate commits) and fixed to work with my other PR. Just needs some work on the tests, then it's ready to go.

chemoish commented:
+1

gkatsev commented on src/js/clickable-component.js:
```diff
@@ -130,6 +130,30 @@ class ClickableComponent extends Component {
+   */
+  enable() {
+    this.removeClass('vjs-disabled');
+    this.el_.setAttribute('aria-disabled', 'false');
```
not necessary for this, but we're doing a lot of attribute setting, we probably should make a helper for components so you could just do `this.setAttribute('aria-disabled', 'false');`

OwenEdwards replied:
:+1: That would be a good place to check that we're consistently passing 'false' or 'true' (strings) to ARIA attributes, rather than booleans, which I *think* causes problems.

gkatsev commented:
Looks like the main missing unit test from this PR is one that checks that the descriptions button is disabled when  caption or subtitle is enabled.
There are also a few unit tests missing for menu-button and clickable-component around enabling/disabling and pressing/unpressing.
I think the descriptions button test should be considered a blocker for this PR the others can always be added later.

OwenEdwards commented:
@gkatsev I think this covers the tests we talked about. Can you take a look and let me know?

gkatsev commented on test/unit/menu.test.js:
```diff
@@ -18,4 +19,44 @@ test('should place title list item into ul', function() {
+q.test('clicking should display the menu', function() {
+  expect(6);
+
+  var player, menuButton, el;
```
should be 3 `let`s.
We really need to update videojs to use `vjsstandard` already so I won't have to keep pointing these out myself :)

gkatsev commented on test/unit/tracks/text-track-controls.test.js:
```diff
@@ -110,6 +110,130 @@ test('menu should update with removeRemoteTextTrack', function() {
   player.dispose();
 });
 
+var descriptionstrack = {
```
`let`.

gkatsev commented on test/unit/tracks/text-track-controls.test.js:
```diff
@@ -110,6 +110,130 @@ test('menu should update with removeRemoteTextTrack', function() {
+});
+
+test('descriptions should be displayed when a text track is added to an empty track list', function() {
+  var player = TestHelpers.makePlayer();
```
let

gkatsev commented on test/unit/tracks/text-track-controls.test.js:
```diff
@@ -110,6 +110,130 @@ test('menu should update with removeRemoteTextTrack', function() {
+});
+
+test('descriptions should not be displayed when text tracks list is empty', function() {
+  var player = TestHelpers.makePlayer();
```
let

gkatsev commented on test/unit/tracks/text-track-controls.test.js:
```diff
@@ -110,6 +110,130 @@ test('menu should update with removeRemoteTextTrack', function() {
+});
+
+test('descriptions should not be displayed when last text track is removed', function() {
+  var player = TestHelpers.makePlayer({
```
let

gkatsev commented on test/unit/tracks/text-track-controls.test.js:
```diff
@@ -110,6 +110,130 @@ test('menu should update with removeRemoteTextTrack', function() {
+});
+
+test('descriptions menu should contain "Off" and one track', function() {
+  var player = TestHelpers.makePlayer({
```
two lets

gkatsev commented on test/unit/tracks/text-track-controls.test.js:
```diff
@@ -110,6 +110,130 @@ test('menu should update with removeRemoteTextTrack', function() {
+test('enabling a captions track should disable the descriptions menu button', function() {
+  expect(14);
+
+  var player = TestHelpers.makePlayer({
```
let

gkatsev commented on test/unit/tracks/text-track-controls.test.js:
```diff
@@ -110,6 +110,130 @@ test('menu should update with removeRemoteTextTrack', function() {
+  ok(!player.controlBar.captionsButton.hasClass('vjs-disabled'), 'captions control is NOT disabled');
+  ok(!player.controlBar.descriptionsButton.hasClass('vjs-disabled'), 'descriptions control is NOT disabled');
+
+  for (i = 0; i < player.textTracks().length; i++) {
```
`i` should be declared as a let here directly

gkatsev commented on test/unit/tracks/text-track-controls.test.js:
```diff
@@ -110,6 +110,130 @@ test('menu should update with removeRemoteTextTrack', function() {
+  ok(!player.controlBar.captionsButton.hasClass('vjs-disabled'), 'captions control is NOT disabled');
+  ok(!player.controlBar.descriptionsButton.hasClass('vjs-disabled'), 'descriptions control is NOT disabled');
+
+  for (i = 0; i < player.textTracks().length; i++) {
```
and here

gkatsev commented on test/unit/tracks/text-track-controls.test.js:
```diff
@@ -110,6 +110,130 @@ test('menu should update with removeRemoteTextTrack', function() {
+  ok(!player.controlBar.captionsButton.hasClass('vjs-disabled'), 'captions control is NOT disabled');
+  ok(player.controlBar.descriptionsButton.hasClass('vjs-disabled'), 'descriptions control IS disabled');
+
+  for (i = 0; i < player.textTracks().length; i++) {
```
and here

gkatsev commented:
@OwenEdwards thanks! Looks good. Just some stylistic issues left :)

OwenEdwards commented:
@gkatsev think that does it!

gkatsev commented:
Awesome. LGTM!

OwenEdwards commented:
@gkatsev realized that I needed to add some strings into en.json. Other than that, this is *ready* :-)

gkatsev commented:
Cool!
Should hopefully get merged/released this week!

OwenEdwards commented:
:+1:

chemoish commented:
Kind of an aside, but does description track support work with `data-setup='{}'` configured?

https://github.com/videojs/video.js/issues/2689#issuecomment-146721304

OwenEdwards commented:
@chemoish this change requires a `data-setup='{ "html5" : { "nativeTextTracks" : false }}'`, because even browsers with native HTML5 track support don't support descriptions tracks. I haven't tested it with Opera, but it pasts tests on Firefox, Safari, Chrome and IE.

@gkatsev, I guess we ought to document this somewhere?

gkatsev commented:
Yes, we need a text tracks guide doc.

Also, maybe we should just always subsume descriptions into emulated mode until browsers that implementing them? We can always do that as a follow-up PR.

surikat1978 commented:
button was hidden.    http://2sub.ru/animation-cartoons/18-bambi-2-bembi-2-2006.html
 Apparently she was hiding a gun when there are no options. And options in chrome physically there. The fox option - in the code page list li, li, li, li. The chromium is only one and it is li captions off. In my browser so.
And I just did a button visible, adding

.vjs-captions-button {display: block important;!}

But the options out there still do not physically in the code.

Chrome = = Yandex Opera

no difference, in my code only :

<div class="vjs-menu" role="presentation"><ul class="vjs-menu-content" role="menu">
<li class="vjs-menu-item vjs-selected" tabindex="-1" role="menuitemcheckbox" aria-live="polite" aria-checked="true">captions off<span class="vjs-control-text">, selected</span></li>
</ul></div>

and all. Here I have copied all the code, you see only one point here.
but the FF, compare

<li aria-live="polite" role="menuitem" tabindex="-1" class="vjs-menu-item vjs-texttrack-settings">captions settings<span class="vjs-control-text">, opens captions settings dialog</span></li>
<li aria-checked="true" aria-live="polite" role="menuitemcheckbox" tabindex="-1" class="vjs-menu-item vjs-selected">captions off<span class="vjs-control-text">, selected</span></li>
<li aria-checked="false" aria-live="polite" role="menuitemcheckbox" tabindex="-1" class="vjs-menu-item">Double<span class="vjs-control-text"> </span></li>
<li aria-checked="false" aria-live="polite" role="menuitemcheckbox" tabindex="-1" class="vjs-menu-item">English<span class="vjs-control-text"> </span></li>
<li aria-checked="false" aria-live="polite" role="menuitemcheckbox" tabindex="-1" class="vjs-menu-item">Russian<span class="vjs-control-text"> </span></li>
